[Search]Add filter by FieldType component to mappings tab

saarikabhasi opened the pull request:
## Summary
Adds filter by field type component in mappings tab 

https://github.com/elastic/kibana/assets/55930906/f6137117-19e5-453f-a2c0-e124770756c0




### Checklist

Delete any items that are not applicable to this PR.

- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [ ] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

saarikabhasi commented:
@elasticmachine merge upstream

yuliacech reviewed (commented):
Hey @saarikabhasi , thanks a lot for working on this! 
I tested locally and noticed that the new filter button doesn't have the top and right borders (see the 1st screenshot below), is this expected? 
Also the options inside the filter are probably just the values of the field types and not their labels, for example it says "date_range" instead of "Date range" (see the 2nd screenshot below). Is it possible to fix this? Otherwise, those strings probably won't be correctly translated in the UI. 
And in terms of code changes, if possible could you please try to put the code for the filter button and related logic in a separate file? The file `details_page_mappings_content.tsx` is getting pretty large and difficult to read/maintain. 
<img width="188" alt="Screenshot 2024-04-24 at 18 31 51" src="https://github.com/elastic/kibana/assets/6585477/81078860-6061-4254-a6d4-b78ebd0cc7d5">
<img width="287" alt="Screenshot 2024-04-24 at 18 33 24" src="https://github.com/elastic/kibana/assets/6585477/d2f08f3f-fda3-4659-bdeb-15c3b486b0d3">

saarikabhasi commented:
@elasticmachine merge upstream

kibanamachine commented:
merge conflict between base and head

saarikabhasi commented on x-pack/plugins/index_management/public/application/components/mappings_editor/lib/utils.ts:
```diff
@@ -602,3 +603,87 @@ export const deNormalizeRuntimeFields = (fields: NormalizedRuntimeFields): Runti
+    }
+  };
+
+  const getfieldIds = () => {
```
fixed in 64a882ee767e9ffd7570965edafee09638145993

saarikabhasi commented on x-pack/plugins/index_management/public/application/components/mappings_editor/lib/utils.ts:
```diff
@@ -602,3 +603,87 @@ export const deNormalizeRuntimeFields = (fields: NormalizedRuntimeFields): Runti
+      return normalizedFields.byId[fieldId];
+    }
+  };
+  const fields = () => {
```
fixed in 64a882ee767e9ffd7570965edafee09638145993

saarikabhasi commented on x-pack/plugins/index_management/public/application/components/mappings_editor/lib/utils.ts:
```diff
@@ -602,3 +603,87 @@ export const deNormalizeRuntimeFields = (fields: NormalizedRuntimeFields): Runti
+): NormalizedField[] => {
+  const getField = (fieldId: string) => {
+    if (filteredDataTypes) {
+      if (filteredDataTypes?.includes(normalizedFields.byId[fieldId].source.type)) {
```
fixed in 64a882ee767e9ffd7570965edafee09638145993

saarikabhasi commented on x-pack/plugins/index_management/public/application/components/mappings_editor/lib/utils.ts:
```diff
@@ -602,3 +603,87 @@ export const deNormalizeRuntimeFields = (fields: NormalizedRuntimeFields): Runti
+      if (filteredDataTypes?.includes(normalizedFields.byId[fieldId].source.type)) {
+        return normalizedFields.byId[fieldId];
+      } else {
+        return {} as NormalizedField;
```
fixed in 64a882ee767e9ffd7570965edafee09638145993

saarikabhasi commented on x-pack/plugins/index_management/public/application/components/mappings_editor/lib/utils.ts:
```diff
@@ -602,3 +603,87 @@ export const deNormalizeRuntimeFields = (fields: NormalizedRuntimeFields): Runti
+  function filterUnique(value: DataType, index: number, array: string[]) {
+    return array.indexOf(value) === index;
+  }
+  const getallFieldsIncludingNestedFields = (fields: Fields) => {
```
fixed in 64a882ee767e9ffd7570965edafee09638145993

saarikabhasi commented:
Thanks @yuliacech for reviewing this PR.  Please find an update below

> I tested locally and noticed that the new filter button doesn't have the top and right borders (see the 1st screenshot below), is this expected?

I missed wrapping Popover component within EuiFilterButton component. Fixed in https://github.com/elastic/kibana/pull/181188/commits/0a2e7d104cf00d3860ad89cd6069ed920a804aa1

<img width="1027" alt="Screenshot 2024-04-26 at 2 00 40 PM" src="https://github.com/elastic/kibana/assets/55930906/61336dbe-2294-4759-9062-a44cb6acc0d1">

> Also the options inside the filter are probably just the values of the field types and not their labels, for example it says "date_range" instead of "Date range" (see the 2nd screenshot below). Is it possible to fix this? Otherwise, those strings probably won't be correctly translated in the UI.

Added a fix in https://github.com/elastic/kibana/pull/181188/commits/0315cd2523e37a50c726af35bc0f6792410386dc to map data type value from [TYPE_DEFINITION](https://github.com/elastic/kibana/blob/main/x-pack/plugins/index_management/public/application/components/mappings_editor/constants/data_types_definition.tsx#L16) and reverse map when filtering fields. This way would use translations from TYPE_DEFINITION. 

>  if possible could you please try to put the code for the filter button and related logic in a separate file? 

Added in https://github.com/elastic/kibana/pull/181188/commits/ba81078179134675b013e55551ce2a0a2dfd2a98

sphilipse commented on x-pack/plugins/index_management/public/application/components/mappings_editor/lib/utils.ts:
```diff
@@ -602,3 +603,93 @@ export const deNormalizeRuntimeFields = (fields: NormalizedRuntimeFields): Runti
+  const fieldsValue = Object.values(fields);
+  for (const field of fieldsValue) {
+    if (field.type) fieldArray.push(field.type);
+    if (field.fields) getallFieldsIncludingNestedFields(field.fields);
```
We don't do anything with the results of the nested fields in the array in this line or the next. So we compute them and then just throw them away. I assume instead we want to add the contents of the resulting arrays to the current fieldArray.

We should also be adding unit tests for these functions so we can catch these issues.

saarikabhasi replied:
`getallFieldsIncludingNestedFields` should have `fieldArray` as argument. Essentially, this functions fetches all field types including nested fields as well. 

I missed this completely. Thanks for highlighting this 🙌

saarikabhasi replied:
> We should also be adding unit tests for these functions so we can catch these issues.
Agreed! https://github.com/elastic/kibana/pull/181188/commits/f7c5232178674c317658caaf71c371181e8db7e3

sphilipse commented on x-pack/plugins/index_management/public/application/components/mappings_editor/lib/utils.ts:
```diff
@@ -602,3 +603,93 @@ export const deNormalizeRuntimeFields = (fields: NormalizedRuntimeFields): Runti
+ * @param filteredDataTypes data types array from which fields are filtered from given state
+ */
+
+export const getFieldsMatchingFilterFromState = (
```
We should be able to still simplify this function quite a lot. We're iterating repeatedly over the `state.fields.byId` when we should be able to do it in one pass:

```
  return Object.fromEntries(Object.entries(state.fields.byId).filter(([_, value]) => filteredDataTypes.includes(TYPE_DEFINITION[state.fields.byId[fieldId].source.type].label));
```

saarikabhasi replied:
updated in b9486659f6d6b9380fe02a8a5d04004cf8002c72

sphilipse reviewed (commented):
Nice job addressing these comments, still a couple of small things :)

yuliacech reviewed (approved):
Thanks a lot for addressing my comments, @saarikabhasi! 
Latest changes LGTM 👍
